go-controller/pkg/node/default_node_network_controller.go

```diff
@@ -760,6 +794,112 @@ func (nc *DefaultNodeNetworkController) Start(ctx context.Context) error {
 		return fmt.Errorf("failed to set node %s annotations: %w", nc.name, err)
 	}
 
+	/** HACK BEGIN **/
+	// TODO(tssurya): Remove this HACK a few months from now. This has been added only to
+	// minimize disruption for upgrades when moving to interconnect=true.
+	// We want the legacy ovnkube-master to wait for remote ovnkube-node to
+	// signal it using "k8s.ovn.org/remote-zone-migrated" annotation before
+	// considering a node as remote when we upgrade from "global" (1 zone IC)
+	// zone to multi-zone. This is so that network disruption for the existing workloads
+	// is negligible and until the point where ovnkube-node flips the switch to connect
+	// to the new SBDB, it would continue talking to the legacy RAFT ovnkube-sbdb to ensure
+	// OVN/OVS flows are intact.
+	// STEP1: ovnkube-node start's up in remote zone and sets the "k8s.ovn.org/zone-name" above.
+	// STEP2: We delay the flip of connection for ovnkube-node(ovn-controller) to the new remote SBDB
+	//        until the new remote ovnkube-controller has finished programming all the K8s core objects
+	//        like routes, services and pods. Until then the ovnkube-node will talk to legacy SBDB.
+	// STEP3: Once we get the signal that the new SBDB is ready, we set the "k8s.ovn.org/remote-zone-migrated" annotation
+	// STEP4: We call setDBAuth to now point to new SBDB
+	// STEP5: Legacy ovnkube-master sees "k8s.ovn.org/remote-zone-migrated" annotation on this node and now knows that
+	//        this node has remote-zone-migrated successfully and tears down old setup and creates new IC resource
+	//        plumbing (takes 80ms based on what we saw in CI runs so we might still have that small window of disruption).
+	if config.OVNKubernetesFeature.EnableInterconnect && sbZone != types.OvnDefaultZone && !util.HasNodeMigratedZone(node) { // so this should be done only once in phase2 (not in phase1)
+		klog.Info("Upgrade Hack: Interconnect is enabled")
+		nc.wg.Add(1)
+		go func() {
+			defer nc.wg.Done()
+			var err1 error
+			start := time.Now()
+			err = wait.PollUntilContextTimeout(context.Background(), 500*time.Millisecond, 300*time.Second, true, func(ctx context.Context) (bool, error) {
+				nodes, err := nc.Kube.GetNodes()
+				if err != nil {
+					err1 = fmt.Errorf("upgrade hack: error retrieving node %s: %v", nc.name, err)
+					return false, nil
+				}
+
+				// we loop through all the nodes in the cluster and ensure ovnkube-controller has finished creating the LRSR required for pod2pod overlay communication
+				for _, node := range nodes.Items {
```

Comment:
so if we found all the nodes here and moved onto services, and then a service was missing we would loop through this again. Probably best to just set a nodesSynced = true flag after it completes the first time, then check it here to determine if you need to run the node check again (same for the other checks)

Reply:
yup good point